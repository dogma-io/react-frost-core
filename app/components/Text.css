.text-field-example {
  margin: 0 10px;
  width: 300px;

  > h4 {
    margin-top: 0;
  }
}

.text-field-examples {
  display: flex;
  flex-wrap: wrap;
}
