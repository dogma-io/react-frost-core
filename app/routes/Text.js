/**
 * @flow
 */

import {CodeBlock, Text, TEXT_ALIGN_RIGHT} from '../../src'
import route from './factory'
import React, {type Element, type Node} from 'react'

type DOC_TYPE = {|
  examples: {[key: string]: Element<*>},
  title: string,
|}

const DOCS: Array<DOC_TYPE> = [
  {
    title: 'State',
    examples: {
      Disabled: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text disabled={true} value="can't touch this" />
        </CodeBlock>
      ),
      Error: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text error={true} />
        </CodeBlock>
      ),
      'Read only': (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text readOnly={true} value="Jane Doe" />
        </CodeBlock>
      ),
    },
  },
  {
    title: 'Options',
    examples: {
      Align: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text align={TEXT_ALIGN_RIGHT} />
        </CodeBlock>
      ),
      Autofocus: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text autoFocus={true} />
        </CodeBlock>
      ),
      Placeholder: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text placeholder="Username" />
        </CodeBlock>
      ),
      'Tab index': (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text tabIndex={-1} />
        </CodeBlock>
      ),
      Value: (
        <CodeBlock language="jsx" maxLineLength={30}>
          <Text value="John Doe" />
        </CodeBlock>
      ),
    },
  },
]

const IMPORTS_CODE = `
import {
  Text,
  TEXT_ALIGN_LEFT,
  TEXT_ALIGN_RIGHT,
} from 'react-frost-core'
`.trim()

const renderExamples = (examples: ?{[key: string]: Element<*>}): Node => {
  if (!examples) {
    return null
  }

  return Object.keys(examples).map((key: string, index: number): Node => {
    return (
      <div className="text-field-example" key={index}>
        <h4>{key}</h4>
        {/* $FlowFixMe - Flow thinks examples could be null/undefined */}
        {examples[key]}
      </div>
    )
  })
}

export default route('Text field', (): Node => {
  return [
    <section key="imports">
      <CodeBlock code={IMPORTS_CODE} demo={false} language="js" />
    </section>,
    <section key="default">
      <h3>Default</h3>
      <div className="text-field-examples">
        <div className="text-field-example">
          <CodeBlock language="jsx" maxLineLength={30}>
            <Text />
          </CodeBlock>
        </div>
      </div>
    </section>,
  ].concat(
    DOCS.map(({examples, title}: DOC_TYPE, index: number): Node => {
      return (
        <section key={index}>
          <h3>{title}</h3>
          <div className="text-field-examples">{renderExamples(examples)}</div>
        </section>
      )
    }),
  )
})
