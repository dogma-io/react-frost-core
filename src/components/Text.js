/**
 * @flow
 */

import ClearSVG from './ClearSVG'
import React, {Component, type Node} from 'react'

export const ALIGN_LEFT: 'left' = 'left'
export const ALIGN_RIGHT: 'right' = 'right'

const PREFIX = 'frost-text'

type ALIGN = typeof ALIGN_LEFT | typeof ALIGN_RIGHT

// eslint-disable-next-line flowtype/require-exact-type
export type PROPS = {
  align?: ?ALIGN,
  className?: ?string,
  disabled?: ?boolean,
  error?: ?boolean,
  inputRef?: (el: ?HTMLInputElement) => void,
  maxLength?: ?number,
  minLength?: ?number,
  onChange?: (value: ?string) => void,
  readOnly?: ?boolean,
  size?: ?number,
  value?: ?string,
}

type State = {|
  animatingClearButtonOut: boolean,
  focused: boolean,
  value?: ?string,
|}

/**
 * Get class name for text component given it's current state
 * @param className - user specified class name
 * @param error - whether or not input has an error
 * @param minLength - minimum length input's value should be
 * @param size - size if input
 * @param value - input value
 * @returns class name for text component
 */
function getClassName(
  className?: ?string,
  error?: ?boolean,
  minLength?: ?number,
  size?: ?number,
  value?: ?string,
): string {
  const classNames = [PREFIX]

  if (className) {
    classNames.push(className)
  }

  if (size) {
    classNames.push(`${PREFIX}-size`)
  }

  if (
    error ||
    (typeof minLength === 'number' &&
      (typeof value !== 'string' || minLength > value.length))
  ) {
    classNames.push(`${PREFIX}-error`)
  }

  return classNames.join(' ')
}

/**
 * Get class name for text component's input
 * @param align - how to align text within input
 * @returns class name for text component's input
 */
function getInputClassName(align?: ?ALIGN): string {
  const classNames = [`${PREFIX}-input`]

  switch (align) {
    case ALIGN_RIGHT:
      classNames.push(`${PREFIX}-align-right`)
  }

  return classNames.join(' ')
}

export default class Text extends Component<PROPS, State> {
  constructor() {
    super(...arguments)

    this.state = {
      animatingClearButtonOut: false,
      focused: false,
      value: this.props.value,
    }
  }

  _handleBlur = () => {
    const {disabled, readOnly} = this.props
    const {animatingClearButtonOut, value} = this.state
    // eslint-disable-next-line flowtype/no-weak-types
    const state: any = {focused: false}

    if (!animatingClearButtonOut && !disabled && !readOnly && value) {
      state.animatingClearButtonOut = true
    }

    this.setState(state)
  }

  _handleChange = (e: SyntheticInputEvent<*>) => {
    this._updateValue(e.target.value)
  }

  _handleClear = () => {
    this._updateValue('')
  }

  _handleClearButtonAnimationEnd = () => {
    const {animatingClearButtonOut, value} = this.state

    // We don't want to do anything if this is the fade in animation.
    if (!animatingClearButtonOut && value) {
      return
    }

    this.setState({animatingClearButtonOut: false})
  }

  _handleFocus = () => {
    this.setState({focused: true})
  }

  _renderClearButton(): Node {
    const {disabled, readOnly} = this.props
    const {animatingClearButtonOut, focused, value} = this.state

    if (
      !animatingClearButtonOut &&
      (disabled || !focused || readOnly || !value)
    ) {
      return null
    }

    const classNames = [
      `${PREFIX}-clear`,
      animatingClearButtonOut ? 'frost-fade-out' : 'frost-fade-in',
    ]

    return (
      <button
        className={classNames.join(' ')}
        onAnimationEnd={this._handleClearButtonAnimationEnd}
        onClick={this._handleClear}
        tabIndex={-1}
      >
        <ClearSVG />
      </button>
    )
  }

  _updateValue(nextValue: ?string) {
    const {maxLength, onChange} = this.props
    const {value} = this.state

    if (nextValue === '') {
      nextValue = null
    } else if (
      typeof nextValue === 'string' &&
      typeof maxLength === 'number' &&
      nextValue.length > maxLength
    ) {
      nextValue = nextValue.slice(0, maxLength)
    }

    // Determine if the value was just cleared, which is used to determine
    // when we should start fading the clear button out.
    const valueCleared = !!value && !nextValue

    this.setState({
      animatingClearButtonOut: valueCleared,
      value: nextValue,
    })

    if (onChange) {
      onChange(nextValue)
    }
  }

  render(): Node {
    const {
      align,
      className,
      error,
      inputRef,
      minLength,
      onChange: _onChange,
      size,
      value: _value,
      ...passThroughProps
    } = this.props

    const {value} = this.state

    return (
      <div className={getClassName(className, error, minLength, size, value)}>
        <input
          className={getInputClassName(align)}
          minLength={minLength}
          onBlur={this._handleBlur}
          onChange={this._handleChange}
          onFocus={this._handleFocus}
          ref={inputRef}
          size={size}
          value={value || ''}
          {...passThroughProps}
        />
        {this._renderClearButton()}
      </div>
    )
  }
}
