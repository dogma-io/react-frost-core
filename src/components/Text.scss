@import '../styles/animations';
@import '../styles/colors';
@import '../styles/typography';

$frost-text-border: 1px solid $frost-color-input-border;
$frost-text-color: $frost-color-input;
$frost-text-disabled-bg: $frost-color-input-disabled-bg;
$frost-text-disabled-border: 1px solid $frost-color-input-disabled-border;
$frost-text-error-border: 1px solid $frost-color-input-error-border;
$frost-text-font-size: $frost-font-m;
$frost-text-focus-border: 1px solid $frost-color-input-focus-border;
$frost-text-height: 35px;
$frost-text-hover-border: 1px solid $frost-color-input-hover-border;
$frost-text-padding: 0 30px 0 8px;

.frost-text {
  display: block;
  position: relative;
  min-width: 175px;

  &.frost-text-error {
    .frost-text-input {
      border: $frost-text-error-border;

      &:hover {
        &:enabled {
          &:read-write {
            &:not(:focus) {
              border: $frost-text-error-border;
            }
          }
        }
      }
    }
  }
}

.frost-text-clear {
  display: inline-block;
  fill: $frost-color-grey-6;
  height: 23px;
  position: absolute;
  right: 5px;
  top: 7px;
  width: 23px;

  &:focus {
    outline: none;
  }
}

.frost-text-input {
  display: inline-block;
  width: 100%;
  height: $frost-text-height;
  padding: $frost-text-padding;
  transition: border .2s ease;
  border: $frost-text-border;
  outline: none;
  color: $frost-text-color;
  font-size: $frost-text-font-size;
  font-weight: 200;

  &.frost-text-align-right {
    text-align: right;
  }

  &:focus {
    border: $frost-text-focus-border;
  }

  &:read-only {
    border: 0;
  }

  // Note: without this Firefox ends up applying the read-only pseudo selector
  // to number inputs which causes them not to get a border.
  &[type='number'] {
    border: 1px solid $frost-color-input-border;
  }

  &:disabled {
    border: $frost-text-disabled-border;
    background-color: $frost-text-disabled-bg;
  }

  &:hover {
    &:enabled {
      &:read-write {
        &:not(:focus) {
          border: $frost-text-hover-border;
        }
      }
    }
  }
}
