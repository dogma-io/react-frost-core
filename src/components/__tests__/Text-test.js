import Text from '../Text'
import {mount} from 'enzyme'
import React from 'react'

describe('Text', () => {
  it('renders as expected without any properties passed in', () => {
    expect(mount(<Text />)).toMatchSnapshot()
  })
})
