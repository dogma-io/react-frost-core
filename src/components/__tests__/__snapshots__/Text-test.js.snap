// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders as expected without any properties passed in 1`] = `
<Text>
  <div
    className="frost-text"
  >
    <input
      className="frost-text-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
  </div>
</Text>
`;
